core/src/main/java/org/jdbi/v3/core/mapper/RowMapper.java

```diff
@@ -30,10 +30,24 @@
      * Map the row the result set is at when passed in. This method should not cause the result
      * set to advance, allow jDBI to do that, please.
      *
-     * @param r the result set being iterated
+     * @param rs the result set being iterated
      * @param ctx the statement context
      * @return the value to return for this row
      * @throws SQLException if anything goes wrong go ahead and let this percolate, jDBI will handle it
      */
-    T map(ResultSet r, StatementContext ctx) throws SQLException;
+    T map(ResultSet rs, StatementContext ctx) throws SQLException;
+
+    /**
+     * Returns a row mapper instance with the details of the result set memoized for performance. Overriding this
+     * method is optional--the default implementation simply returns {@code this}.
+     *
+     * @param rs  the result set to memoize over
+     * @param ctx the statement context to memoize over
+     * @return a row mapper equivalent to this one, possibly memoized.
+     * @throws SQLException if anything goes wrong go ahead and let this percolate, jDBI will handle it
+     * @see org.jdbi.v3.core.mapper.reflect.BeanMapper for an example of memoization.
+     */
+    default RowMapper<T> memoize(ResultSet rs, StatementContext ctx) throws SQLException {
```

Comment:
I don't think we can, or should, pull this up to the factories. There's an implicit order that we find out what the user wants:
* What type they want to map to (before the query executes)
* What columns are available to map (after the query executes)

Existing methods like `Query.map(RowMapper)` are called before the query executes, so unless we want to take away that convenience, this is a nonstarter.

I did a spike on my previous comment, (see the `jdbi3-row-suppliers` branch) and ended up with this:

```java
interface RowMapperFactory<T> {
  Optional<RowMapper<T>> build(Type type, ConfigRegistry config);
}

interface RowMapper<T> {
  RowSupplier<T> bind(ResultSet rs, StatementContext ctx) throws SQLException;
}

interface RowSupplier<T> {
  T get() throws SQLException;
}
```

This approach meant I had to convert lambda mappers in tests to higher order e.g. `(rs, ctx) -> rs.getString("foo")`  became `(rs, ctx) -> () -> rs.getString("foo")`. In one case a test broke without any compiler errors, because the existing lambda returned null. That's a papercut I'd rather not introduce.

So my next idea will be to revert `RowMapper` change, rename it, and and introduce a new, parallel mapping pipeline. The new pipeline will be the "official" pipeline and the old mapper will be adapted transparently to work in the new pipeline:

```java
interface RowBinderFactory<T> {
  Optional<RowBinder<T>> build(Type type, ConfigRegistry config);
}

interface RowBinder<T> {
  RowSupplier<T> bind(ResultSet rs, StatementContext ctx) throws SQLException;
}

interface RowSupplier<T> {
  T get() throws SQLException;
}

interface RowMapper<T> { // adapted into RowBinder when registered
  T map(ResultSet rs, StatementContext ctx) throws SQLException;
}
```

These new class names are just working names--I'm open to suggestions.

One possible drawback is that you can mix up `Query.registerRowBinder` and `registerRowMapper` and use a lambda, it's possible the compiler won't catch it because the two types take the same arguments.

Thoughts?